embedding_reader/parquet_numpy_reader.py

```diff
@@ -62,9 +73,14 @@ def __call__(self, batch_size, start=0, end=None, max_piece_size=None, parallel_
             batch_size = end - start
 
         if max_piece_size is None:
-            max_piece_size = max(int(50 * 10**6 / (self.byte_per_item)), 1)
+            # Take x embeddings per pieces so that the max piece size is max(50MB, size of a batch)
+            max_piece_size = max(int(50 * 10**6 / self.byte_per_item), batch_size, 1)
+
         if parallel_pieces is None:
-            parallel_pieces = max(math.ceil(batch_size / max_piece_size), 10)
+            # read_piece function can load the whole file in memory to take only max_piece_size bytes
+            read_piece_ram_usage = self.average_file_size
```

Comment:
Thanks for the review!

Indeed, I did like that because I wanted to avoid the edge case when there is a single big file constraining the whole reader to use only one thread to read the embeddings while it could use many threads for the other files (a best-effort constraint rather than a hard one).

But yeah, even for standard case (we files are stored at a constant size and the last file is smaller than the others) the average would have been a wrong metric. The median would have been better at least

I will go for the max solution as it is the one that has all the guaranties 👍 

(A cleaner solution would be to create a variable quantity of threads dynamically so that the total memory used is always below the constraint, but let's keep things simple for now)

Reply:
You are right, we could probably enter into the edge case even in a standard case.

Let's go for max in the first place like you said.

Reply:
Good catch. One possible way to address this (for another review) would be to have a ThreadPool to the max size we define (50 somewhere in the code), and then to offer treatments to the thread pool always in a manner that the memory is not over the cap. I am not sure how to do that on python but I am thinking about a dedicated iterator to provide to imap that keeps track of 'in progress reads', blocks if offering any new embedding to read would overcome the memory threshold, otherwise offers any read that could fit in.
Another approach is to prevent a reader thread to start if it would consume more memory than what is accepted by keeping a shared memory count that is increased/decrease by readers method entry/exit. I tend to prefer acting at the moment jobs are offered to threads rather than in the thread itself, but all of that depends on the primitive available in python